clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -448,40 +448,48 @@ public ReadyCheckResult ready(Cluster cluster, long nowMs) {
         boolean exhausted = this.free.queued() > 0;
         for (Map.Entry<TopicPartition, Deque<ProducerBatch>> entry : this.batches.entrySet()) {
             Deque<ProducerBatch> deque = entry.getValue();
+
+            final ProducerBatch batch;
+            final long waitedTimeMs;
+            final boolean backingOff;
+            final boolean full;
+
+            // Collect as little as possible inside critical region, determine outcome after release
             synchronized (deque) {
-                // When producing to a large number of partitions, this path is hot and deques are often empty.
-                // We check whether a batch exists first to avoid the more expensive checks whenever possible.
```

Comment:
Why did we remove the information on why this is performance sensitive?

Reply:
I attempted to capture the sentiment with this comment, given the re-flowed logic:
```
// Collect as little as possible inside critical region, determine outcome after release
```
I'll add something more descriptive.

Reply:
I think the useful context that was removed is:

> When producing to a large number of partitions, this path is hot and deques are often empty.

Reply:
[f15fdef](https://github.com/apache/kafka/pull/11722/commits/f15fdef3241a17f86af44c6dac04ba651d629426)

Added:

```
// This loop is especially hot with large partition counts.

// We are careful to only perform the minimum required inside the
// synchronized block, as this lock is also used to synchronize producer threads
// attempting to append() to a partition/batch.

synchronized (deque) {
    // Deques are often empty in this path, esp with large partition counts,
    // so we exit early if we can.
```